src/execution/operator/join/physical_asof_join.cpp

```diff
@@ -392,273 +489,325 @@ void AsOfLocalState::ResolveJoin(DataChunk &input, bool *found_match, std::pair<
 
-	AsOfGlobalSinkState &gstate;
+	AsOfGlobalSourceState &gsource;
+
+	//! The left side partition being probed
+	AsOfProbeBuffer probe_buffer;
 
 	//! The read partition
 	idx_t hash_bin;
 	HashGroupPtr hash_group;
-
 	//! The read cursor
 	unique_ptr<PayloadScanner> scanner;
-	//! Buffer for the inputs
-	DataChunk input_chunk;
 	//! Pointer to the matches
 	const bool *found_match;
 };
 
-AsOfLocalSourceState::AsOfLocalSourceState(AsOfGlobalSinkState &gstate_p) : gstate(gstate_p) {
-	input_chunk.Initialize(gstate.global_partition.allocator, gstate.global_partition.payload_types);
+AsOfLocalSourceState::AsOfLocalSourceState(AsOfGlobalSourceState &gsource, const PhysicalAsOfJoin &op)
+    : gsource(gsource), probe_buffer(gsource.gsink.lhs_sink->context, op) {
+	gsource.mergers++;
+}
+
+void AsOfLocalSourceState::CombineLeftPartitions() {
+	const auto buffer_count = gsource.gsink.lhs_buffers.size();
+	while (gsource.combined < buffer_count) {
+		const auto next_combine = gsource.next_combine++;
+		if (next_combine < buffer_count) {
+			gsource.gsink.lhs_buffers[next_combine]->Combine();
+			++gsource.combined;
+		} else {
+			std::this_thread::yield();
+		}
+	}
+}
+
+void AsOfLocalSourceState::MergeLeftPartitions() {
+	PartitionGlobalMergeStates::Callback local_callback;
+	PartitionLocalMergeState local_merge;
+	gsource.GetMergeStates().ExecuteTask(local_merge, local_callback);
+	gsource.merged++;
+	while (gsource.merged < gsource.mergers) {
+		std::this_thread::yield();
+	}
 }
 
-idx_t AsOfLocalSourceState::GeneratePartition(const idx_t hash_bin_p) {
-	//	Get rid of any stale data
+idx_t AsOfLocalSourceState::BeginRightScan(const idx_t hash_bin_p) {
 	hash_bin = hash_bin_p;
 
-	hash_group = std::move(gstate.global_partition.hash_groups[hash_bin]);
+	hash_group = std::move(gsource.gsink.rhs_sink.hash_groups[hash_bin]);
 	scanner = make_uniq<PayloadScanner>(*hash_group->global_sort);
-	found_match = gstate.right_outers[hash_bin].GetMatches();
+	found_match = gsource.gsink.right_outers[hash_bin].GetMatches();
 
 	return scanner->Remaining();
 }
 
 unique_ptr<LocalSourceState> PhysicalAsOfJoin::GetLocalSourceState(ExecutionContext &context,
                                                                    GlobalSourceState &gstate) const {
-	auto &gsink = sink_state->Cast<AsOfGlobalSinkState>();
-	return make_uniq<AsOfLocalSourceState>(gsink);
+	auto &gsource = gstate.Cast<AsOfGlobalSourceState>();
+	return make_uniq<AsOfLocalSourceState>(gsource, *this);
 }
 
 SourceResultType PhysicalAsOfJoin::GetData(ExecutionContext &context, DataChunk &chunk,
                                            OperatorSourceInput &input) const {
-	D_ASSERT(IsRightOuterJoin(join_type));
-
 	auto &gsource = input.global_state.Cast<AsOfGlobalSourceState>();
 	auto &lsource = input.local_state.Cast<AsOfLocalSourceState>();
-	auto &gsink = gsource.gsink;
+	auto &rhs_sink = gsource.gsink.rhs_sink;
+
+	//	Step 1: Combine the partitions
+	lsource.CombineLeftPartitions();
```

Comment:
I thought about making a double pipeline breaker, but `ExecuteInternal` returns left values that cannot pass the predicates because they are `NULL` so this needs to be done in the source.